Introduce `-Zembed-metadata` to allow omitting full metadata from rlibs and dylibs

Kobzol opened the pull request:
This is a continuation of https://github.com/rust-lang/rust/pull/120855 (I was mentored by @bjorn3 to move it forward). Most of the original code was written by bjorn3, I tried to clean it up a bit and add some documentation and tests.

This PR introduces a new unstable compiler flag called `-Zembed-metadata=[no|yes]`, with the default being `yes` (see https://github.com/rust-lang/rust/issues/57076 for context). When set to `no`, rustc will only store a small metadata stub inside rlibs/dylibs instead of the full metadata, to keep their size smaller. It should be used in combination with `--emit=metadata`, so that the users of such a compiled library can still read the metadata from the corresponding `.rmeta` file. [This comment](https://github.com/rust-lang/rust/pull/120855#issuecomment-1937018169) shows an example of binary/artifact size wins that can be achieved using this approach.

Contrary to https://github.com/rust-lang/rust/pull/120855, this PR only introduces the new flag, along with a couple of run-make tests and documentation, but does not yet use it in bootstrap to actually compile rustc. I plan to do that as a follow-up step (along with integration in Cargo, which should ideally just always pass this flag to reduce the size of target directories).

Fixes https://github.com/rust-lang/rust/issues/23366
Closes https://github.com/rust-lang/rust/issues/29511
Fixes https://github.com/rust-lang/rust/issues/57076

Another attempt of https://github.com/rust-lang/rust/pull/93945 and https://github.com/rust-lang/rust/pull/120855.

r? @petrochenkov

bjorn3 committed:
Simplify find_commandline_library

Kobzol committed:
Add `-Zembed-metadata` CLI option

Kobzol committed:
Add documentation of the option into the unstable book

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13497004436/job/37706450175) [(plain)](https://github.com/rust-lang/rust/commit/bf5a7c2bc04b7514df4db13a8817c4bde7cd412e/checks/37706450175/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 29.7s done
#21 DONE 35.9s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 4.31s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
warning: target feature `x87` must be enabled to ensure that the ABI of the current target can be implemented correctly
  |
  = note: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: for more information, see issue #116344 <https://github.com/rust-lang/rust/issues/116344>
---
---- [run-make] tests/run-make/split-metadata stdout ----

error: rmake recipe failed to complete
status: exit status: 1
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/split-metadata/rmake_out" && env -u RUSTFLAGS AR="ar" BUILD_ROOT="/checkout/obj/build/x86_64-unknown-linux-gnu" CARGO="/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/cargo" CC="cc" CC_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" CXX="c++" CXX_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" HOST_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" LD_LIB_PATH_ENVVAR="LD_LIBRARY_PATH" LLVM_BIN_DIR="/usr/lib/llvm-18/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgputargetmca amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen codegentypes core coroutines coverage debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerclassic dwarflinkerparallel dwp engine executionengine extensions filecheck frontenddriver frontendhlsl frontendoffloading frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo hipstdpar instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcdebugging orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi textapibinaryreader transformutils ve veasmparser vecodegen vectorize vedesc vedisassembler veinfo webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" LLVM_FILECHECK="/usr/lib/llvm-18/bin/FileCheck" NODE="/usr/bin/node" PYTHON="/usr/bin/python3" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTDOC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" SOURCE_ROOT="/checkout" TARGET="x86_64-unknown-linux-gnu" TARGET_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" TARGET_RPATH_ENV="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/split-metadata/rmake_out:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/split-metadata/rmake"
--- stderr -------------------------------
Testing library kind Rlib
command failed at line 43
command failed at line 43
Command { cmd: LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/split-metadata/temporary-directory:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/split-metadata/temporary-directory" "foo.rs", stdin_buf: None, stdin: None, stdout: None, stderr: None, drop_bomb: DropBomb { command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc", defused: true, armed_location: Location { file: "/checkout/tests/run-make/split-metadata/rmake.rs", line: 43, col: 9 } }, already_executed: true }
output status: `exit status: 101`
=== STDOUT ===


=== STDERR ===

---
   8:     0x7f4fa806af1b - std[1b36a9b3e2321cd0]::panicking::begin_panic_handler::{closure#0}
   9:     0x7f4fa80681a9 - std[1b36a9b3e2321cd0]::sys::backtrace::__rust_end_short_backtrace::<std[1b36a9b3e2321cd0]::panicking::begin_panic_handler::{closure#0}, !>
  10:     0x7f4fa806ab35 - rust_begin_unwind
  11:     0x7f4fa3eed532 - core[87d8de562df86e0e]::panicking::panic_fmt
  12:     0x7f4fa3ee9e46 - <rustc_serialize[2a0ca79c5c179d6e]::opaque::MemDecoder>::decoder_exhausted
  13:     0x7f4fa71d18b7 - rustc_serialize[2a0ca79c5c179d6e]::leb128::read_usize_leb128
  14:     0x7f4fa71d3323 - <rustc_metadata[ea0efcbd17c2d30a]::rmeta::decoder::DecodeContext as rustc_serialize[2a0ca79c5c179d6e]::serialize::Decoder>::read_str
  15:     0x7f4fa71d61cd - <rustc_metadata[ea0efcbd17c2d30a]::rmeta::CrateRoot as rustc_serialize[2a0ca79c5c179d6e]::serialize::Decodable<rustc_metadata[ea0efcbd17c2d30a]::rmeta::decoder::DecodeContext>>::decode
  16:     0x7f4fa72158f6 - <rustc_metadata[ea0efcbd17c2d30a]::rmeta::decoder::MetadataBlob>::get_root
  17:     0x7f4fa7155fd1 - <rustc_metadata[ea0efcbd17c2d30a]::creader::CrateLoader>::load
  18:     0x7f4fa7150d16 - <rustc_metadata[ea0efcbd17c2d30a]::creader::CrateLoader>::maybe_resolve_crate
  19:     0x7f4fa714f5c7 - <rustc_metadata[ea0efcbd17c2d30a]::creader::CrateLoader>::resolve_crate
  20:     0x7f4fa7159d60 - <rustc_metadata[ea0efcbd17c2d30a]::creader::CrateLoader>::process_extern_crate
  21:     0x7f4fa5d19a38 - <rustc_resolve[7cc96f8fc6cdbab8]::Resolver>::crate_loader::<core[87d8de562df86e0e]::option::Option<rustc_span[dc91d7b7a49d219c]::def_id::CrateNum>, <rustc_resolve[7cc96f8fc6cdbab8]::build_reduced_graph::BuildReducedGraphVisitor>::build_reduced_graph_for_extern_crate::{closure#0}>
  22:     0x7f4fa5e40a0d - <rustc_resolve[7cc96f8fc6cdbab8]::build_reduced_graph::BuildReducedGraphVisitor as rustc_ast[a500c1de1b33acbb]::visit::Visitor>::visit_item
  23:     0x7f4fa5e43bce - <rustc_resolve[7cc96f8fc6cdbab8]::build_reduced_graph::BuildReducedGraphVisitor as rustc_ast[a500c1de1b33acbb]::visit::Visitor>::visit_crate
  24:     0x7f4fa5d06e55 - <rustc_resolve[7cc96f8fc6cdbab8]::Resolver as rustc_expand[ecf4876e26faa4ee]::base::ResolverExpand>::visit_ast_fragment_with_placeholders
  25:     0x7f4fa7311704 - <rustc_expand[ecf4876e26faa4ee]::expand::MacroExpander>::collect_invocations
  26:     0x7f4fa730b3ce - <rustc_expand[ecf4876e26faa4ee]::expand::MacroExpander>::fully_expand_fragment
  27:     0x7f4fa730afd2 - <rustc_expand[ecf4876e26faa4ee]::expand::MacroExpander>::expand_crate
  28:     0x7f4fa42da009 - <rustc_session[abf0c519ca9e2e2b]::session::Session>::time::<rustc_ast[a500c1de1b33acbb]::ast::Crate, rustc_interface[e3342892a47d9ab8]::passes::configure_and_expand::{closure#1}>
  29:     0x7f4fa41cf2b1 - rustc_interface[e3342892a47d9ab8]::passes::resolver_for_lowering_raw
  30:     0x7f4fa6441ef8 - rustc_query_impl[2a9e3d5c3a9147ad]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[2a9e3d5c3a9147ad]::query_impl::resolver_for_lowering_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[b8f1cd3a173958f4]::query::erase::Erased<[u8; 16usize]>>
  31:     0x7f4fa63acc09 - <rustc_query_impl[2a9e3d5c3a9147ad]::query_impl::resolver_for_lowering_raw::dynamic_query::{closure#2} as core[87d8de562df86e0e]::ops::function::FnOnce<(rustc_middle[b8f1cd3a173958f4]::ty::context::TyCtxt, ())>>::call_once
  32:     0x7f4fa63392b8 - <std[1b36a9b3e2321cd0]::thread::local::LocalKey<core[87d8de562df86e0e]::cell::Cell<*const ()>>>::with::<rustc_middle[b8f1cd3a173958f4]::ty::context::tls::enter_context<rustc_query_system[86acc13612db55b]::query::plumbing::execute_job_non_incr<rustc_query_impl[2a9e3d5c3a9147ad]::DynamicConfig<rustc_query_system[86acc13612db55b]::query::caches::SingleCache<rustc_middle[b8f1cd3a173958f4]::query::erase::Erased<[u8; 16usize]>>, false, false, false>, rustc_query_impl[2a9e3d5c3a9147ad]::plumbing::QueryCtxt>::{closure#0}, rustc_middle[b8f1cd3a173958f4]::query::erase::Erased<[u8; 16usize]>>::{closure#0}, rustc_middle[b8f1cd3a173958f4]::query::erase::Erased<[u8; 16usize]>>
  33:     0x7f4fa626def1 - rustc_query_system[86acc13612db55b]::query::plumbing::try_execute_query::<rustc_query_impl[2a9e3d5c3a9147ad]::DynamicConfig<rustc_query_system[86acc13612db55b]::query::caches::SingleCache<rustc_middle[b8f1cd3a173958f4]::query::erase::Erased<[u8; 16usize]>>, false, false, false>, rustc_query_impl[2a9e3d5c3a9147ad]::plumbing::QueryCtxt, false>
  34:     0x7f4fa6604cec - rustc_query_impl[2a9e3d5c3a9147ad]::query_impl::resolver_for_lowering_raw::get_query_non_incr::__rust_end_short_backtrace
  35:     0x7f4fa77e8147 - <rustc_middle[b8f1cd3a173958f4]::ty::context::TyCtxt>::resolver_for_lowering
  36:     0x7f4fa3ff91eb - <std[1b36a9b3e2321cd0]::thread::local::LocalKey<core[87d8de562df86e0e]::cell::Cell<*const ()>>>::with::<rustc_middle[b8f1cd3a173958f4]::ty::context::tls::enter_context<<rustc_middle[b8f1cd3a173958f4]::ty::context::GlobalCtxt>::enter<rustc_interface[e3342892a47d9ab8]::passes::create_and_enter_global_ctxt<core[87d8de562df86e0e]::option::Option<rustc_interface[e3342892a47d9ab8]::queries::Linker>, rustc_driver_impl[1019500eb6a6c9a3]::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}, core[87d8de562df86e0e]::option::Option<rustc_interface[e3342892a47d9ab8]::queries::Linker>>::{closure#1}, core[87d8de562df86e0e]::option::Option<rustc_interface[e3342892a47d9ab8]::queries::Linker>>::{closure#0}, core[87d8de562df86e0e]::option::Option<rustc_interface[e3342892a47d9ab8]::queries::Linker>>
  37:     0x7f4fa3f8edc0 - <rustc_middle[b8f1cd3a173958f4]::ty::context::TyCtxt>::create_global_ctxt::<core[87d8de562df86e0e]::option::Option<rustc_interface[e3342892a47d9ab8]::queries::Linker>, rustc_interface[e3342892a47d9ab8]::passes::create_and_enter_global_ctxt<core[87d8de562df86e0e]::option::Option<rustc_interface[e3342892a47d9ab8]::queries::Linker>, rustc_driver_impl[1019500eb6a6c9a3]::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}>
  38:     0x7f4fa3f8cf14 - <rustc_interface[e3342892a47d9ab8]::passes::create_and_enter_global_ctxt<core[87d8de562df86e0e]::option::Option<rustc_interface[e3342892a47d9ab8]::queries::Linker>, rustc_driver_impl[1019500eb6a6c9a3]::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core[87d8de562df86e0e]::ops::function::FnOnce<(&rustc_session[abf0c519ca9e2e2b]::session::Session, rustc_middle[b8f1cd3a173958f4]::ty::context::CurrentGcx, &std[1b36a9b3e2321cd0]::sync::once_lock::OnceLock<rustc_middle[b8f1cd3a173958f4]::ty::context::GlobalCtxt>, &rustc_data_structures[a66926230b4ccfed]::sync::worker_local::WorkerLocal<rustc_middle[b8f1cd3a173958f4]::arena::Arena>, &rustc_data_structures[a66926230b4ccfed]::sync::worker_local::WorkerLocal<rustc_hir[44d5bee511119f4]::Arena>, rustc_driver_impl[1019500eb6a6c9a3]::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  39:     0x7f4fa3fd455b - <alloc[b0929b49c7791218]::boxed::Box<dyn for<'a> core[87d8de562df86e0e]::ops::function::FnOnce<(&'a rustc_session[abf0c519ca9e2e2b]::session::Session, rustc_middle[b8f1cd3a173958f4]::ty::context::CurrentGcx, &'a std[1b36a9b3e2321cd0]::sync::once_lock::OnceLock<rustc_middle[b8f1cd3a173958f4]::ty::context::GlobalCtxt<'a>>, &'a rustc_data_structures[a66926230b4ccfed]::sync::worker_local::WorkerLocal<rustc_middle[b8f1cd3a173958f4]::arena::Arena<'a>>, &'a rustc_data_structures[a66926230b4ccfed]::sync::worker_local::WorkerLocal<rustc_hir[44d5bee511119f4]::Arena<'a>>, rustc_driver_impl[1019500eb6a6c9a3]::run_compiler::{closure#0}::{closure#2}), Output = core[87d8de562df86e0e]::option::Option<rustc_interface[e3342892a47d9ab8]::queries::Linker>>> as core[87d8de562df86e0e]::ops::function::FnOnce<(&rustc_session[abf0c519ca9e2e2b]::session::Session, rustc_middle[b8f1cd3a173958f4]::ty::context::CurrentGcx, &std[1b36a9b3e2321cd0]::sync::once_lock::OnceLock<rustc_middle[b8f1cd3a173958f4]::ty::context::GlobalCtxt>, &rustc_data_structures[a66926230b4ccfed]::sync::worker_local::WorkerLocal<rustc_middle[b8f1cd3a173958f4]::arena::Arena>, &rustc_data_structures[a66926230b4ccfed]::sync::worker_local::WorkerLocal<rustc_hir[44d5bee511119f4]::Arena>, rustc_driver_impl[1019500eb6a6c9a3]::run_compiler::{closure#0}::{closure#2})>>::call_once
  40:     0x7f4fa3f8b77d - rustc_interface[e3342892a47d9ab8]::passes::create_and_enter_global_ctxt::<core[87d8de562df86e0e]::option::Option<rustc_interface[e3342892a47d9ab8]::queries::Linker>, rustc_driver_impl[1019500eb6a6c9a3]::run_compiler::{closure#0}::{closure#2}>
  41:     0x7f4fa3fbda34 - <scoped_tls[4c63abcec0c49958]::ScopedKey<rustc_span[dc91d7b7a49d219c]::SessionGlobals>>::set::<rustc_interface[e3342892a47d9ab8]::util::run_in_thread_with_globals<rustc_interface[e3342892a47d9ab8]::util::run_in_thread_pool_with_globals<rustc_interface[e3342892a47d9ab8]::interface::run_compiler<(), rustc_driver_impl[1019500eb6a6c9a3]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}, ()>
  42:     0x7f4fa3ff825c - rustc_span[dc91d7b7a49d219c]::create_session_globals_then::<(), rustc_interface[e3342892a47d9ab8]::util::run_in_thread_with_globals<rustc_interface[e3342892a47d9ab8]::util::run_in_thread_pool_with_globals<rustc_interface[e3342892a47d9ab8]::interface::run_compiler<(), rustc_driver_impl[1019500eb6a6c9a3]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}>
  43:     0x7f4fa3f648d1 - std[1b36a9b3e2321cd0]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[e3342892a47d9ab8]::util::run_in_thread_with_globals<rustc_interface[e3342892a47d9ab8]::util::run_in_thread_pool_with_globals<rustc_interface[e3342892a47d9ab8]::interface::run_compiler<(), rustc_driver_impl[1019500eb6a6c9a3]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  44:     0x7f4fa3ff5c50 - <<std[1b36a9b3e2321cd0]::thread::Builder>::spawn_unchecked_<rustc_interface[e3342892a47d9ab8]::util::run_in_thread_with_globals<rustc_interface[e3342892a47d9ab8]::util::run_in_thread_pool_with_globals<rustc_interface[e3342892a47d9ab8]::interface::run_compiler<(), rustc_driver_impl[1019500eb6a6c9a3]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[87d8de562df86e0e]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  46:     0x7f4fa328aaa4 - <unknown>
  47:     0x7f4fa3317c3c - <unknown>
  48:                0x0 - <unknown>

```

</details>

Kobzol commented:
(The test failure is the one situation not yet implemented, which is described in the PR description - `--extern` plus `.rmeta` in a different directory than the rlib/dylib).

petrochenkov commented:
>I found it quite surprising that when `--extern` is passed, the `-L` search logic is completely skipped

That's by design, `--extern` with paths disables all the directory search, makes the build system own the input files, and makes rustc just do what the build system tells it.
I'd actually consider this the primary mode of work for rustc, and all the directory search is a sort of legacy (well, except `-L dependency=PATH`).

`--extern` supports passing multiple paths for the same crate, to specify all its rmeta, rlib and dylib versions.
So in case of split metadata I'd also expect cargo (or another build system) to pass path to the split rmeta explicitly, and rustc to not do any heuristics.

In #120855 the heuristics were a temporary measure to enable split metadata in bootstrap, while cargo still doesn't support it.
It should be fine to add them in your follow up PR implementing the bootstrap changes as well, but in the long run I don't think they should exist.

Kobzol commented:
Interesting. So if I understood it correctly, you would like to remove all heuristics from rustc (so no lookup of `.rmeta` in known library search dirs and no ".rmeta in the same directory as .rlib" heuristic) and let Cargo pass essentially both `.rlib` and `.rmeta` for every dependency through `--extern`? I guess that's cleaner in a sense.

On the other hand, it would essentially double the length of the rustc command line, because for every library that you depend on, Cargo would have to pass both `.rmeta` and `.rlink`. Do you think that wouldn't be an issue? I suppose that it would only happen for the leaf linked artifact, because the intermediate libraries seem to only receive `.rmeta` files from Cargo, while the final artifact receives `.rlib`.

petrochenkov commented:
>So if I understood it correctly, you would like to remove all heuristics from rustc...

Yes.

>On the other hand, it would essentially double the length of the rustc command line, because for every library that you depend on, Cargo would have to pass both `.rmeta` and `.rlib`.

My impression was that it is already the case with pipelined builds, but apparently it's indeed only `.rmeta`s for dependencies and only `.rlib`s for the root crate.
I think it's fine if it happens only for direct dependencies of binary crates, in any case we can reconsider heuristics if there are any complaints about this.

Kobzol committed:
Add an error when full metadata was not found

Kobzol committed:
Add tests

Kobzol commented:
Ok, rebased and removed the heuristics. I added an error message when the full metadata is not found, but I'm not sure if we should emit the error eagerly (left a comment in code). Could you please take a look at https://github.com/rust-lang/rust/pull/137535/commits/fe29fec36695c6e9f0fdb29ae74b0789782eda74? Thank you!

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13854961190/job/38769701307) [(plain)](https://github.com/rust-lang/rust/commit/c878f22b4daa298fcd238edcc628877cd7ab0791/checks/38769701307/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 2.798 Building wheels for collected packages: reuse
#12 2.799   Building wheel for reuse (pyproject.toml): started
#12 3.018   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.019   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=0c2fd2aaf7b0bf8d6e131220aff14712a774c2ca462f3204d25460cbcf610b63
#12 3.019   Stored in directory: /tmp/pip-ephem-wheel-cache-p640v34w/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.021 Successfully built reuse
#12 3.022 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.429 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.429 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 3.975 Collecting virtualenv
#12 4.012   Downloading virtualenv-20.29.3-py3-none-any.whl (4.3 MB)
#12 4.095      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 53.7 MB/s eta 0:00:00
#12 4.136 Collecting distlib<1,>=0.3.7
#12 4.140   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.147      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 84.9 MB/s eta 0:00:00
#12 4.186 Collecting filelock<4,>=3.12.2
#12 4.190   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.223 Collecting platformdirs<5,>=3.9.1
#12 4.227   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#12 4.310 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.494 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.6 virtualenv-20.29.3
#12 4.494 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 4.6s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      124864 kB
DirectMap2M:     5117952 kB
DirectMap1G:    13631488 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
[TIMING] core::build_steps::tool::LibcxxVersionTool { target: x86_64-unknown-linux-gnu } -- 0.220
---
[TIMING] core::build_steps::tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
fmt: checked 5908 files
tidy check
tidy error: /checkout/compiler/rustc_metadata/messages.ftl: message `metadata_full_metadata_not_found` appears before `metadata_fail_create_file_encoder`, but is alphabetically later than it
run `./x.py test tidy --bless` to sort the file correctly
tidy: Skipping binary file check, read-only filesystem
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.0.1)
linting python files
All checks passed!
checking python file formatting
26 files already formatted
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:49
  local time: Fri Mar 14 10:56:30 UTC 2025
  network time: Fri, 14 Mar 2025 10:56:30 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bjorn3 commented on compiler/rustc_metadata/src/locator.rs:
```diff
@@ -654,7 +654,22 @@ impl<'a> CrateLocator<'a> {
-            *slot = Some((hash, metadata, lib.clone()));
+
+            // Question: maybe we shouldn't error eagerly here, because it is possible that
+            // e.g. a rlib only contains a stub, but a (later-resolved) dylib contains the full
```
`-Zsplit-metadata` applies to all outputs. And both rlib and dylib must be compiled by the same rustc invocation to be guaranteed that the crate hash matches (and thus they are considered interchangeable in dependency resolution).

Kobzol replied:
Ok, that makes sense. Thanks!

Kobzol commented:
https://github.com/rust-lang/compiler-team/issues/851 was accepted, so I think that we can go forward with this.

@petrochenkov What do you think about renaming the flag to `-Zembed-metadata=[no|yes]`, with `yes` being the default, to be consistent with `-Cembed-bitcode=[no|yes]`, which is a bit similar in spirit?

petrochenkov commented:
> What do you think about renaming the flag to `-Zembed-metadata=[no|yes]`, with `yes` being the default, to be consistent with `-Cembed-bitcode=[no|yes]`, which is a bit similar in spirit?

Seems ok to me.

Kobzol committed:
Store only a metadata stub into `rlibs` and `dylibs` with `-Zembed-metadata=no`

Kobzol commented:
Ok, changed it to `-Zembed-metadata=[no|yes]`.

@rustbot ready

jieyouxu commented on tests/run-make/embed-metadata/rmake.rs:
Suggestion: could you back link to the tracking issue?

bjorn3 replied:
This is a new unstable feature, so a tracking issue should be opened for it.

Kobzol replied:
Ah, right, good point! Created it here: https://github.com/rust-lang/rust/issues/139165.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14134233019/job/39602045768) [(plain)](https://github.com/rust-lang/rust/commit/5d072bb9fbac4b7826ebc73ecaa445c13039d84f/checks/39602045768/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#19 exporting to docker image format
#19 sending tarball 20.9s done
#19 DONE 35.5s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Number of decisions:   4447
  longest path:          1159 (code:    152)
  longest backtrack:       66 (code:    428)
Shared 86733 out of 152951 states by creating 14756 new states, saving 71977
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc: In function ‘rtx_def* extract_bit_field_1(rtx, poly_uint64, poly_uint64, int, rtx, machine_mode, machine_mode, bool, bool, rtx_def**)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc:1864:45: warning: ‘*(unsigned int*)((char*)&imode + offsetof(scalar_int_mode, scalar_int_mode::m_mode))’ may be used uninitialized [-Wmaybe-uninitialized]
 1864 |       rtx sub = extract_bit_field_as_subreg (mode1, op0, imode,
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
 1865 |                                              bitsize, bitnum);
      |                                              ~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc:1824:19: note: ‘*(unsigned int*)((char*)&imode + offsetof(scalar_int_mode, scalar_int_mode::m_mode))’ was declared here
 1824 |   scalar_int_mode imode;
      |                   ^~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gimple-range-gori.cc: In member function ‘void range_def_chain::dump(FILE*, basic_block, const char*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gimple-range-gori.cc:319:19: warning: format not a string literal and no format arguments [-Wformat-security]
  319 |           fprintf (f, prefix);
      |           ~~~~~~~~^~~~~~~~~~~
---
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:7930:9: warning: ignoring return value of ‘ssize_t write(int, const void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 7930 |   write (fd, "\n\n", 2);
      |   ~~~~~~^~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc: In member function ‘void driver::final_actions() const’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:9307:13: warning: ignoring return value of ‘int truncate(const char*, __off_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 9307 |     truncate(totruncate_file, 0);
      |     ~~~~~~~~^~~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto-wrapper.cc: In function ‘bool find_and_merge_options(int, off_t, const char*, vec<cl_decoded_option>, bool, vec<cl_decoded_option>*, const char*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto-wrapper.cc:1165:8: warning: ignoring return value of ‘ssize_t read(int, void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 1165 |   read (fd, data, length);
      |   ~~~~~^~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc: In function ‘void lto_resolution_read(splay_tree, FILE*, lto_file*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2091:10: warning: ignoring return value of ‘int fscanf(FILE*, const char*, ...)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2091 |   fscanf (resolution, " ");   /* Read white space.  */
      |   ~~~~~~~^~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2093:9: warning: ignoring return value of ‘size_t fread(void*, size_t, size_t, FILE*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2093 |   fread (obj_name, sizeof (char), name_len, resolution);
      |   ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2113:10: warning: ignoring return value of ‘int fscanf(FILE*, const char*, ...)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2113 |   fscanf (resolution, "%u", &num_symbols);
      |   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.cc:32:
---
Applying io_quotes_use            to linux/blkzoned.h
Applying io_quotes_use            to linux/ipmi.h
Applying io_quotes_use            to linux/psp-dbc.h
Applying io_quotes_use            to linux/bt-bmc.h
Applying io_quotes_use            to linux/tps6594_pfsm.h
Applying io_quotes_use            to linux/cxl_mem.h
Applying io_quotes_use            to linux/wmi.h
Applying io_quotes_use            to linux/auto_fs.h
Applying io_quotes_use            to linux/mmtimer.h
Applying io_quotes_use            to linux/f2fs.h
Applying io_quotes_use            to linux/vhost.h
---
Applying machine_name             to x86_64-linux-gnu/bits/unistd_ext.h
Applying io_quotes_use            to x86_64-linux-gnu/asm/mtrr.h
Applying io_quotes_use            to x86_64-linux-gnu/asm/amd_hsmp.h
Applying machine_name             to openssl/e_os2.h
Applying io_quotes_use            to drm/xe_drm.h
Applying io_quotes_use            to drm/radeon_drm.h
Applying io_quotes_use            to drm/panfrost_drm.h
Applying io_quotes_use            to drm/etnaviv_drm.h
Applying io_quotes_use            to drm/lima_drm.h
Applying io_quotes_use            to drm/qaic_accel.h
Applying io_quotes_use            to drm/vc4_drm.h
Applying io_quotes_use            to drm/i915_drm.h
Applying io_quotes_use            to drm/omap_drm.h
Applying io_quotes_use            to drm/pvr_drm.h
Applying io_quotes_use            to drm/amdgpu_drm.h
Applying io_quotes_use            to drm/vgem_drm.h
Applying io_quotes_use            to drm/msm_drm.h
Applying io_quotes_use            to drm/v3d_drm.h
Applying io_quotes_use            to drm/exynos_drm.h
Applying io_quotes_use            to drm/nouveau_drm.h
Applying io_quotes_use            to drm/drm.h
Applying io_quotes_use            to drm/habanalabs_accel.h
Applying io_quotes_use            to drm/tegra_drm.h
Applying io_quotes_use            to rdma/rdma_user_ioctl.h
cc1: note: self-tests are not enabled in this build
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/c++tools/server.cc: In function ‘void server(bool, int, module_resolver*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/c++tools/server.cc:620:10: warning: ignoring return value of ‘int pipe(int*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
---
[RUSTC-TIMING] proc_macro test:false 5.247
[RUSTC-TIMING] test test:false 9.107
    Finished `release` profile [optimized] target(s) in 1m 20s
##[endgroup]
[TIMING] core::build_steps::compile::Std { target: x86_64-unknown-linux-gnu, compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu, forced_compiler: true }, crates: [], force_recompile: false, extra_rust_args: ["-Csymbol-mangling-version=v0", "-Cpanic=abort"], is_for_mir_opt_tests: false } -- 80.496
Testing GCC stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
 Downloading crates ...
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
[RUSTC-TIMING] boml test:false 0.813
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
[RUSTC-TIMING] y test:false 3.026
    Finished `release` profile [optimized] target(s) in 4.15s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-backend gcc --gcc-path /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/install/lib --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
`--gcc-path` was provided, ignoring config file. Using `/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/install/lib` as path for libgccjit
[BUILD] mini_core
[RUSTC-TIMING] mini_core test:false 0.197
[BUILD] example
[AOT] mini_core_hello_world
[RUSTC-TIMING] mini_core_hello_world test:false 0.186
```

</details>

Kobzol commented:
Created a tracking issue and fixed tests.

petrochenkov commented on compiler/rustc_metadata/src/rmeta/mod.rs:
```diff
@@ -221,6 +221,12 @@ pub(crate) struct CrateHeader {
+    /// in a separate rmeta file).
+    ///
+    /// This is used inside rlibs and dylibs when using `-Zembed-metadata=no`.
+    pub(crate) is_stub: bool,
```
The comment above says "If you do modify this struct, also bump the [`METADATA_VERSION`] constant."

Kobzol replied:
Oh, good catch, thanks!

petrochenkov commented:
r=me with the metadata version updated.
@rustbot author

Kobzol committed:
Bump metadata version

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14188298745/job/39747499732) [(plain)](https://github.com/rust-lang/rust/commit/cb7ebf6f05a08a3dc7d2562eed1a3e3c5523ff3a/checks/39747499732/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#19 exporting to docker image format
#19 sending tarball 19.7s done
#19 DONE 30.7s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Number of decisions:   4447
  longest path:          1159 (code:    152)
  longest backtrack:       66 (code:    428)
Shared 86733 out of 152951 states by creating 14756 new states, saving 71977
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc: In function ‘rtx_def* extract_bit_field_1(rtx, poly_uint64, poly_uint64, int, rtx, machine_mode, machine_mode, bool, bool, rtx_def**)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc:1864:45: warning: ‘*(unsigned int*)((char*)&imode + offsetof(scalar_int_mode, scalar_int_mode::m_mode))’ may be used uninitialized [-Wmaybe-uninitialized]
 1864 |       rtx sub = extract_bit_field_as_subreg (mode1, op0, imode,
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
 1865 |                                              bitsize, bitnum);
      |                                              ~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc:1824:19: note: ‘*(unsigned int*)((char*)&imode + offsetof(scalar_int_mode, scalar_int_mode::m_mode))’ was declared here
 1824 |   scalar_int_mode imode;
      |                   ^~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gimple-range-gori.cc: In member function ‘void range_def_chain::dump(FILE*, basic_block, const char*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gimple-range-gori.cc:319:19: warning: format not a string literal and no format arguments [-Wformat-security]
  319 |           fprintf (f, prefix);
      |           ~~~~~~~~^~~~~~~~~~~
---
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:7930:9: warning: ignoring return value of ‘ssize_t write(int, const void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 7930 |   write (fd, "\n\n", 2);
      |   ~~~~~~^~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc: In member function ‘void driver::final_actions() const’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:9307:13: warning: ignoring return value of ‘int truncate(const char*, __off_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 9307 |     truncate(totruncate_file, 0);
      |     ~~~~~~~~^~~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto-wrapper.cc: In function ‘bool find_and_merge_options(int, off_t, const char*, vec<cl_decoded_option>, bool, vec<cl_decoded_option>*, const char*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto-wrapper.cc:1165:8: warning: ignoring return value of ‘ssize_t read(int, void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 1165 |   read (fd, data, length);
      |   ~~~~~^~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc: In function ‘void lto_resolution_read(splay_tree, FILE*, lto_file*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2091:10: warning: ignoring return value of ‘int fscanf(FILE*, const char*, ...)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2091 |   fscanf (resolution, " ");   /* Read white space.  */
      |   ~~~~~~~^~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2093:9: warning: ignoring return value of ‘size_t fread(void*, size_t, size_t, FILE*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2093 |   fread (obj_name, sizeof (char), name_len, resolution);
      |   ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2113:10: warning: ignoring return value of ‘int fscanf(FILE*, const char*, ...)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2113 |   fscanf (resolution, "%u", &num_symbols);
      |   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.cc:32:
---
Applying io_quotes_use            to linux/blkzoned.h
Applying io_quotes_use            to linux/ipmi.h
Applying io_quotes_use            to linux/psp-dbc.h
Applying io_quotes_use            to linux/bt-bmc.h
Applying io_quotes_use            to linux/tps6594_pfsm.h
Applying io_quotes_use            to linux/cxl_mem.h
Applying io_quotes_use            to linux/wmi.h
Applying io_quotes_use            to linux/auto_fs.h
Applying io_quotes_use            to linux/mmtimer.h
Applying io_quotes_use            to linux/f2fs.h
Applying io_quotes_use            to linux/vhost.h
---
Applying machine_name             to x86_64-linux-gnu/bits/unistd_ext.h
Applying io_quotes_use            to x86_64-linux-gnu/asm/mtrr.h
Applying io_quotes_use            to x86_64-linux-gnu/asm/amd_hsmp.h
Applying machine_name             to openssl/e_os2.h
Applying io_quotes_use            to drm/xe_drm.h
Applying io_quotes_use            to drm/radeon_drm.h
Applying io_quotes_use            to drm/panfrost_drm.h
Applying io_quotes_use            to drm/etnaviv_drm.h
Applying io_quotes_use            to drm/lima_drm.h
Applying io_quotes_use            to drm/qaic_accel.h
Applying io_quotes_use            to drm/vc4_drm.h
Applying io_quotes_use            to drm/i915_drm.h
Applying io_quotes_use            to drm/omap_drm.h
Applying io_quotes_use            to drm/pvr_drm.h
Applying io_quotes_use            to drm/amdgpu_drm.h
Applying io_quotes_use            to drm/vgem_drm.h
Applying io_quotes_use            to drm/msm_drm.h
Applying io_quotes_use            to drm/v3d_drm.h
Applying io_quotes_use            to drm/exynos_drm.h
Applying io_quotes_use            to drm/nouveau_drm.h
Applying io_quotes_use            to drm/drm.h
Applying io_quotes_use            to drm/habanalabs_accel.h
Applying io_quotes_use            to drm/tegra_drm.h
Applying io_quotes_use            to rdma/rdma_user_ioctl.h
cc1: note: self-tests are not enabled in this build
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/c++tools/server.cc: In function ‘void server(bool, int, module_resolver*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/c++tools/server.cc:620:10: warning: ignoring return value of ‘int pipe(int*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
---
[RUSTC-TIMING] proc_macro test:false 5.162
[RUSTC-TIMING] test test:false 9.403
    Finished `release` profile [optimized] target(s) in 1m 18s
##[endgroup]
[TIMING] core::build_steps::compile::Std { target: x86_64-unknown-linux-gnu, compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu, forced_compiler: true }, crates: [], force_recompile: false, extra_rust_args: ["-Csymbol-mangling-version=v0", "-Cpanic=abort"], is_for_mir_opt_tests: false } -- 78.708
Testing GCC stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
 Downloading crates ...
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
[RUSTC-TIMING] boml test:false 0.805
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
[RUSTC-TIMING] y test:false 2.990
    Finished `release` profile [optimized] target(s) in 4.11s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-backend gcc --gcc-path /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/install/lib --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
`--gcc-path` was provided, ignoring config file. Using `/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/install/lib` as path for libgccjit
[BUILD] mini_core
[RUSTC-TIMING] mini_core test:false 0.194
[BUILD] example
[AOT] mini_core_hello_world
[RUSTC-TIMING] mini_core_hello_world test:false 0.183
---
failures:

---- tests::test_derive_empty stdout ----

thread 'tests::test_derive_empty' panicked at src/tools/rust-analyzer/crates/proc-macro-srv/src/tests/utils.rs:58:48:
called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unsupported metadata version 10" })

---- tests::list_test_macros stdout ----

thread 'tests::list_test_macros' panicked at src/tools/rust-analyzer/crates/proc-macro-srv/src/tests/utils.rs:111:44:
called `Result::unwrap()` on an `Err` value: "Cannot create expander for /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/build/proc-macro-test-10a8964eabd7b9a0/out/target/debug/libproc_macro_test_impl.so: unsupported metadata version 10"
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- tests::test_derive_error stdout ----

thread 'tests::test_derive_error' panicked at src/tools/rust-analyzer/crates/proc-macro-srv/src/tests/utils.rs:58:48:
called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unsupported metadata version 10" })

---- tests::test_fn_like_fn_like_span_ops stdout ----

thread 'tests::test_fn_like_fn_like_span_ops' panicked at src/tools/rust-analyzer/crates/proc-macro-srv/src/tests/utils.rs:58:48:
called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unsupported metadata version 10" })

---- tests::test_fn_like_macro_clone_ident_subtree stdout ----

thread 'tests::test_fn_like_macro_clone_ident_subtree' panicked at src/tools/rust-analyzer/crates/proc-macro-srv/src/tests/utils.rs:58:48:
called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unsupported metadata version 10" })

---- tests::test_attr_macro stdout ----

thread 'tests::test_attr_macro' panicked at src/tools/rust-analyzer/crates/proc-macro-srv/src/tests/utils.rs:58:48:
called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unsupported metadata version 10" })

---- tests::test_fn_like_fn_like_span_join stdout ----

thread 'tests::test_fn_like_fn_like_span_join' panicked at src/tools/rust-analyzer/crates/proc-macro-srv/src/tests/utils.rs:58:48:
called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unsupported metadata version 10" })

---- tests::test_fn_like_macro_clone_literals stdout ----

thread 'tests::test_fn_like_macro_clone_literals' panicked at src/tools/rust-analyzer/crates/proc-macro-srv/src/tests/utils.rs:58:48:
called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unsupported metadata version 10" })

---- tests::test_fn_like_macro_clone_raw_ident stdout ----

thread 'tests::test_fn_like_macro_clone_raw_ident' panicked at src/tools/rust-analyzer/crates/proc-macro-srv/src/tests/utils.rs:58:48:
called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unsupported metadata version 10" })

---- tests::test_fn_like_macro_noop stdout ----

thread 'tests::test_fn_like_macro_noop' panicked at src/tools/rust-analyzer/crates/proc-macro-srv/src/tests/utils.rs:58:48:
called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unsupported metadata version 10" })

---- tests::test_fn_like_mk_literals stdout ----

thread 'tests::test_fn_like_mk_literals' panicked at src/tools/rust-analyzer/crates/proc-macro-srv/src/tests/utils.rs:58:48:
called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unsupported metadata version 10" })

---- tests::test_fn_like_mk_idents stdout ----

thread 'tests::test_fn_like_mk_idents' panicked at src/tools/rust-analyzer/crates/proc-macro-srv/src/tests/utils.rs:58:48:
called `Result::unwrap()` on an `Err` value: Io(Custom { kind: InvalidData, error: "unsupported metadata version 10" })

---- dylib::version::test_version_check stdout ----

thread 'dylib::version::test_version_check' panicked at src/tools/rust-analyzer/crates/proc-macro-srv/src/dylib/version.rs:156:6:
called `Result::unwrap()` on an `Err` value: Custom { kind: InvalidData, error: "unsupported metadata version 10" }


failures:
    tests::test_derive_empty
    tests::list_test_macros
```

</details>

Kobzol committed:
Support metadata version 10 in proc-macro-srv

Kobzol commented:
@bors r=petrochenkov rollup=never

Marked as rollup=never, because this modifies metadata version.

bors commented:
:pushpin: Commit f0efb9748ea02b7e6f15bef45881403ddcfba125 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov f0efb9748ea02b7e6f15bef45881403ddcfba125 -->
<!-- homu: {"type":"Approved","sha":"f0efb9748ea02b7e6f15bef45881403ddcfba125","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit f0efb9748ea02b7e6f15bef45881403ddcfba125 with merge 8c35f4a85e2c505d22c751ebdc292e851b1584cf...
<!-- homu: {"type":"BuildStarted","head_sha":"f0efb9748ea02b7e6f15bef45881403ddcfba125","merge_sha":"8c35f4a85e2c505d22c751ebdc292e851b1584cf"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14193540162/job/39775142523)
Approved by: petrochenkov
Pushing 8c35f4a85e2c505d22c751ebdc292e851b1584cf to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14193540162/job/39775142523"},"merge_sha":"8c35f4a85e2c505d22c751ebdc292e851b1584cf"} -->
